provider.go

```diff
@@ -42,12 +48,51 @@ type ClientProvider struct {
 //     gate := g8.NewGate(g8.NewAuthorizationService().WithClientProvider(clientProvider))
 func NewClientProvider(getClientByTokenFunc func(token string) *Client) *ClientProvider {
 	return &ClientProvider{
-		cache:                false,
 		getClientByTokenFunc: getClientByTokenFunc,
 	}
 }
 
+// WithCache adds cache options to the ClientProvider
+// ttl is the time until the cache entry will be deleted. A ttl of -1 means no expiration
+// maxSize is the maximum amount of entries that can be in the cache at any given time. If a value of 0 or less is provided, it means
+// infinite
+//
+// Example:
+// 		 clientProvider := g8.NewClientProvider(func(token string) *g8.Client {
+//         // We'll assume that the following function calls your database and returns a struct "User" that
+//         // has the user's token as well as the permissions granted to said user
+//         user := database.GetUserByToken(token)
+//         if user != nil {
+//             return g8.NewClient(user.Token).WithPermissions(user.Permissions)
+//         }
+//         return nil
+// 			})
+// 			clientProvider.WithCache(60*time.Minute, 70000)
+//     gate := g8.NewGate(g8.NewAuthorizationService().WithClientProvider(clientProvider))
+func (provider *ClientProvider) WithCache(ttl time.Duration, maxSize int) *ClientProvider {
+	provider.cache = gocache.NewCache().WithEvictionPolicy(gocache.LeastRecentlyUsed).WithMaxSize(maxSize)
+	provider.cache.StartJanitor() // Passively manage expired entries
+
+	provider.ttl = ttl
+	return provider
+}
+
 // GetClientByToken retrieves a client by its token through the provided getClientByTokenFunc.
 func (provider *ClientProvider) GetClientByToken(token string) *Client {
```

Comment:
I think this function could be rewritten as:
```go
// GetClientByToken retrieves a client by its token through the provided getClientByTokenFunc.
func (provider *ClientProvider) GetClientByToken(token string) *Client {
	// No need to go further if cache isn't enabled
	if provider.cache == nil {
		return provider.getClientByTokenFunc(token)
	}

	if value, exists := provider.cache.Get(token); exists {
		return value.(*Client)
	}

        // Client value isn't cached yet
	client := provider.getClientByTokenFunc(token)
	if client != nil {
		provider.cache.SetWithTTL(token, client, provider.ttl)
		return client
	}

	return nil
}
```

I find it more idiomatic.

Reply:
A cache entry could have a `nil` client, therefore, this could cause a panic:
```go
	if value, exists := provider.cache.Get(token); exists {
		return value.(*Client)
	}
```